packages/gatsby/src/utils/cache.ts

```diff
@@ -0,0 +1,77 @@
+import manager, { Store, StoreConfig, CachingConfig } from "cache-manager"
+import fs from "fs-extra"
+import fsStore from "cache-manager-fs-hash"
+import path from "path"
+
+const MAX_CACHE_SIZE = 250
+const TTL = Number.MAX_SAFE_INTEGER
+
+interface ICacheProperties {
+  name?: string
+  store?: Store
+}
+
+class Cache {
+  public name: string
+  public store: Store
+  public cache?: manager.Cache
+
+  constructor({ name = `db`, store = fsStore }: ICacheProperties = {}) {
+    this.name = name
+    this.store = store
+  }
+
+  get directory(): string {
+    return path.join(process.cwd(), `.cache/caches/${this.name}`)
+  }
+
+  init(): Cache {
+    fs.ensureDirSync(this.directory)
+
+    const configs: StoreConfig[] = [
+      {
+        store: `memory`,
+        max: MAX_CACHE_SIZE,
+        ttl: TTL,
+      },
+      {
+        store: this.store,
+        ttl: TTL,
+        options: {
+          path: this.directory,
+          ttl: TTL,
+        },
+      },
+    ]
+
+    const caches = configs.map(cache => manager.caching(cache))
+
+    this.cache = manager.multiCaching(caches)
+
+    return this
+  }
+
+  get<T = unknown>(key): Promise<T | undefined> {
+    return new Promise(resolve => {
+      // eslint-disable-next-line no-unused-expressions
+      this.cache?.get<T>(key, (err, res) => {
+        resolve(err ? undefined : res)
+      })
+    })
+  }
+
+  set<T>(
+    key: string,
+    value: T,
+    args: CachingConfig = { ttl: TTL }
+  ): Promise<T | undefined> {
+    return new Promise(resolve => {
+      // eslint-disable-next-line no-unused-expressions
+      this.cache?.set(key, value, args, err => {
+        resolve(err ? undefined : value)
```

Comment:
Can we use `git mv` to preserve git history? It's hard to review this file because I can't tell what's different. E.g., this resolve(undefined) on error feels like a bug.

Reply:
I'm a bit surpised it didn't pick up that this was a rename. That resolve(undefined) was in the original file.